docs/user-guide/pydantic-schema-manager.rst

```diff
@@ -0,0 +1,226 @@
+
+.. code-block:: python
+
+   from safir.schema_manager import (
+       IncompatibleSchemaError,
+       SchemaRegistryCompatibility,
+       PydanticSchemaManager,
+   )
+
+   manager: PydanticSchemaManager
+
+
+   class MyModel(AvroBaseModel):
+       field1: int
+       field2: str
+
+
+   await manager.register(
+       MyModel, compatibility=SchemaManagerCompatibility.FORWARD
+   )
+
+.. code-block:: python
+
+   manager: PydanticSchemaManager
+
+
+   # sometime in the future, this model changes like this
+   class MyModel(AvroBaseModel):
+       field1: int
+
+
+   # This will throw an exception!
+   await manager.register(MyModel)
+
+.. _compatibility type: https://docs.confluent.io/platform/current/schema-registry/fundamentals/schema-evolution.html#compatibility-types
+
+Subject names
+-------------
+
+The `subject`_ that a schema is registered under is completely independent of any Kafka topics that serialized messages may or may not be published to.
+In other words, it uses the `RecordNameStrategy`_.
+The manager uses the combined Avro namespace and record name as the subject name.
+The record name and namespace come from the name of the model class, and/or certain fields on an inner class named ``Meta``:
+
+* ``schema_name``
+* ``namespace``
+
+.. code-block:: python
+
+   # subject: "MyModel"
+   class MyModel(AvroBaseModel):
+       str_field: str
+       int_field: int
+
+
+.. code-block:: python
+
+   # subject: "mymodelcustom"
+   class MyModel(AvroBaseModel):
+       str_field: str
+       int_field: int
+
+       class Meta:
+           schema_name = "mymodelcustom"
+
+.. code-block:: python
+
+   # subject: my.namespace.mymodelcustom
+   class MyModel(AvroBaseModel):
+       str_field: str
+       int_field: int
+
+       class Meta:
+           schema_name = "mymodelcustom"
+           namespace = "my.namespace"
+
+.. code-block:: python
+
+   # subject: my.namespace.MyModel
+   class MyModel(AvroBaseModel):
+       str_field: str
+       int_field: int
+
+       class Meta:
+           namespace = "my.namespace"
+
+.. _subject: https://docs.confluent.io/platform/current/schema-registry/fundamentals/index.html#schemas-subjects-and-topics
+.. _RecordNameStrategy: https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#sr-schemas-subject-name-strategy
+
+Subject suffixes for development
+================================
+
+When you're developing and testing your app, you probably don't want to register new versions of its schemas in the subjects that actual deployed versions of the app are using.
+You can instantiate the ``PydanticSchemaManager`` with a ``suffix`` argument to add that suffix onto all subjects used by the manager:
+
+.. code-block:: python
+
+   registry: AsyncSchemaRegistryClient
+   manager = PydanticSchemaManager(registry=registry, suffix="_testing")
```

Comment:
I don't think it makes sense to tie it to `config.profile`, assuming our standard practice is to set that to "development" in the Phalanx dev environments, and "production" everywhere else.  In that case, there would still be the same subject for any model deployed to one of those environments in that environment's Schema Registry, it would just be different than that model's subject in the non-dev environments' schema managers. You could still clobber the compatibility in a `config.profile = development` environment by not setting a _different_ subject when you point your phalanx app to a different branch that has model changes that don't make it back into the main branch.

If we wanted to do something automatic, I think we'd want something like, in dev envs, the suffix is the branch name.  In that case though, whatever is on the other end, presumably a different phalanx app, would have to deal with all of the different subjects.

In reality, I think we're likely to screw up compatibility in the registry in a dev env, and we can "just" delete the schema or specific versions of it from the registry when that happens. It stinks, and I don't immediately know how to do any better than "here's a think you can use to try to avoid that if you can remember it" :(

I think that pain is partly why @jonathansick moved away from using the Schema Registry (and to shared Pydantic models) when we control both ends, but I don't want to put words in his mouth, and he may have some more useful thoughts than me on this.

Reply:
I'm totally wrong, `suffix` has no effect on the Avro record name, and I don't think anyone consuming the encoded messages cares about what subject they're in in the Schema Registry.

Reply:
Ah, okay, yeah, `config.profile` doesn't seem to make sense.

I think I'm just confused about how I should decide when to add a prefix and when to remove it again and the process seems somewhat error-prone. It sounds like that's somewhat inherent in the problem space, though, and we may not be able to avoid it.

Could it make sense to use a compatibility setting for the dev environment that allows people to push entirely incompatible schemas, and then use a tighter one in the integration environment? And then really stress in the documentation that you must finalize your schema before you deploy on integration?

Reply:
> And then really stress in the documentation that you must finalize your schema before you deploy on integration?

This for sure!

> Could it make sense to use a compatibility setting for the dev environment that allows people to push entirely incompatible schemas, and then use a tighter one in the integration environment?

I was initially against this because it would be nice to find out if your schemas are incompatible before you get to int (and thus likely after your changes have been merged), but in a previous comment I said "we're likely to screw it up in dev and have to delete subjects and versions", and if we did that, then it's not any better than if we had the NONE compatibility mode set at the registry level.

But another issue with this is that folks's code could manually be setting the compatibility mode differently for different subjects (the EventManager will do this), so if folks use this `PydanticSchemaManager directly, they'll have to explicitly write the code that says "if config.profile is development, set the subject compatibility to NONE".

It would be nice if there was something we can do in a CI run to check compatibility against prod or int, what do you think of doing something like that?

Reply:
Ok, like we talked about in co-work, we're going to [kick a more robust solution to this down the road](https://rubinobs.atlassian.net/browse/DM-46529).  For now, I put a "don't screw this up" warning in the docs.

The metrics EventManager will set all subjects to be `FORWARD` compatible, which has been stable with EFD schema management usage. Any screw ups will be fixed manually for now :crossed_fingers: